### DISCUSSION TITLE
Public ojs javascript api

### DISCUSSION THREAD

**Discussion by anthonynorth** [quarto-dev/quarto-cli#3142] [Q&A] [answered]
Hi,

Is there a public javascript api for ojs inputs and variables? I would like to subscribe to changes to ojs inputs in other on-page javascript. I can do this via the `_ojs` exposed global, but its name suggests private api.

Thanks

**Comment by cscheid** [comment, marked as answer]
We don't explicitly support conversion between OJS and plain Javascript, because the two execution models don't perfectly match.

My first suggestion would be to convert the on-page javascript itself to OJS. If that's not possible, though, you should be able to do that conversion yourself in OJS. It'll require a bit of knowledge of Observable's internal execution model, but consider something like:

````
```{ojs}
//| echo: false
//| display: false
// let reactiveVar be an OJS reactive
{
    window.yourCallback(reactiveVar); // `yourCallback` is a field of the global object.
    return null;
}
```
````

I haven't tested this, but it should work. If you use `reactiveVar` in an OJS block, you get a reactive, and then `window.yourCallback` is called every time `reactiveVar` changes.

**Reply by anthonynorth** [reply]
Thanks. I thought directly using _ojs was a hack. It was a very easy hack though :)

My javscript is a htmlwidget in this case, so my api for defining ojs dependencies will be via R. I should be able to generate an ojs chunk with something similar to your example. Is there an api for this?

**Reply by cscheid** [reply]
No, unfortunately there isn't in this case. I would just use the _ojs "internal" as a workaround until we design a real API around it.

(Something like this is coming, but likely will be in dev for 1.4, so at least 3 months from now.)

**Reply by anthonynorth** [reply]
Thanks.
